### PR TITLE
6532-Incomplete-state-of-method-AST

### PR DESCRIPTION
**Pull request by MarcusDenker**
This is a try to cache the method as part of the AST. There is a strange failure in my image, let's see if this happens on the CI, too

### CODE REVIEW

**Comment by MarcusDenker** [comment]
MustBeBooleanTest>>testWhile is failing, somehow the bc->AST mapping is out of sync

File: src/OpalCompiler-Core/RBMethodNode.extension.st [hunk 1]
```diff
@@ -62,6 +64,7 @@ RBMethodNode >> generate: trailer [
 { #category : #'*OpalCompiler-Core' }
 RBMethodNode >> generateIR [
 	| ir |
+	bcToASTCache := nil.
```
Comment (dionisiydk) [inline src/OpalCompiler-Core/RBMethodNode.extension.st]: This part is suspicious related to failing test

File: src/AST-Core/ASTCacheMissStrategy.class.st [hunk 2]
```diff
@@ -6,5 +6,8 @@ Class {
 
 { #category : #accessing }
 ASTCacheMissStrategy >> getASTFor: aCompiledMethod [
-	^ aCompiledMethod parseTree doSemanticAnalysisIn: aCompiledMethod methodClass
+	| ast |
+	ast := aCompiledMethod parseTree doSemanticAnalysisIn: aCompiledMethod methodClass.
```
Comment (dionisiydk) [inline src/AST-Core/ASTCacheMissStrategy.class.st]: it means that "aCompiledMethod parseTree" will miss the compiled method. 
I think it should be set in #parseTree method. This way you will also avoid duplication in reflective method.

Comment (MarcusDenker) [reply src/AST-Core/ASTCacheMissStrategy.class.st]: ah.. I will try that!

**Comment by MarcusDenker** [comment]
maybe the reason is that we have to call sourceNodeForPC: on the block, it is:


sourceNodeForPC: anInteger 
	self methodNode ir.
	(self hasProperty: #ir) ifTrue: [ "FullBlockClosure"
		^ self bcToASTCache nodeForPC: anInteger].
	^ self methodNode sourceNodeForPC: anInteger

File: src/OpalCompiler-Core/RBMethodNode.extension.st (hunk 1 above)
Comment (MarcusDenker) [reply src/OpalCompiler-Core/RBMethodNode.extension.st]: The reset I think is good: if we re-generate the IR, the cache is (potentially) not valid.

What I want to do after this PR is to refactor to *never* cache the IR... that should simplify things a bit

**Comment by MarcusDenker** [comment]
(It is too hot and I start to mix up PRs already.. the last one was for the other strange bug :-)

**Comment by MarcusDenker** [comment]
failing tests no related

**Comment by MarcusDenker** [comment]
Interesting: CI run passed in 40 minutes. That seems much faster then what we saw without this PR...

**Pull request merged**